specs/latest/2.0/index.html

```diff
@@ -1092,6 +1095,63 @@ <h2 class="no-toc">Table of contents</h2>
         it is the responsibility of the WebGL API to ensure that data are accessed consistently. This applies
         even if the buffer is currently bound to a transform feedback binding point.
       </dd>
+      <dt>
+        <p class="idl-code">Promise&lt;ArrayBuffer&gt; getBufferSubDataAsync(GLenum target, GLintptr offset, ArrayBuffer returnedData, unsigned long returnedOffset = 0, unsigned long returnedSize = 0)
+        </p>
+      </dt>
+      <dd>
+        Let <code>buffer</code> be the WebGLBuffer bound to the passed
+        target. Read <code>numBytes</code> bytes from <code>buffer</code> starting at byte
+        offset <code>offset</code> and write them to <code>returnedData</code> starting at byte
+        offset <code>returnedOffset</code>. <code>numBytes</code> is computed via the following
+        algorithm:
+
+        <ul>
+        <li> If <code>returnedSize</code> is zero, <code>numBytes = returnedData.byteLength - returnedOffset</code></li>
+        <li> Otherwise, <code>numBytes = returnedSize</code>.
+        </ul>
+
+        This command executes serially with respect to others in the command stream, but the data is
+        only written to <code>returnedData</code> when the promise is fulfilled; it is available in
+        the Promise's success callback, but not earlier.
+
+        <br><br>
+        If a failure occurs that prevents the promise from being fulfilled, such as:
+
+        <ul>
+          <li>no WebGLBuffer is bound to <code>target</code></li>
+          <li><code>offset</code> is less than zero</li>
+          <li><code>offset + numBytes</code> would extend beyond the end of <code>buffer</code></li>
+          <li><code>returnedOffset + numBytes</code> would extend beyond the end of <code>returnedData</code>
+          <li><code>target</code> is <code>TRANSFORM_FEEDBACK_BUFFER</code>, and any transform feedback object
+            is currently active</li>
+          <li>the context is lost before the Promise is fulfilled</li>
+        <ul>
+
+        <br><br>
+
+        then the returned Promise is rejected with an "InvalidStateError" DOMException, and no data
+        is written to <code>returnedData</code>.
+
+        <br><br>
+
+        Otherwise, the Promise is fulfilled when the data becomes available,
+        and <code>returnedData</code>, originally passed to <code>getBufferSubDataAsync</code>, is
+        passed to the success callback.
+
+        <br><br>
+        If the buffer is written and read sequentially by other operations
+        and <code>getBufferSubDataAsync</code>, it is the responsibility of the WebGL API to ensure
+        that data are accessed consistently. This applies even if the buffer is currently bound to a
+        transform feedback binding point.
+
```

Comment:
Implementation question: Suppose a web developer gives an ArrayBuffer to this API and then turns around and reads and writes to the buffer.  When the GPU thread is fulfilling the request, I presume it is going to fulfill it into its own separate buffer and then copy the separate buffer into the JavaScript buffer right before fulfilling the promise?  Otherwise, how else are we going to avoid partially available data before the promise is fulfilled?  I presume this is what you mean by "it is the responsibility of the WebGL API to ensure the data are accessed consistently"?  

Along a similar line, if the WebGL developer uses the same buffer for two calls to the async API, it is guaranteed that the promise for the first call will be fulfilled before the promise for the second call?